hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieLogFileReader.java

```diff
@@ -243,6 +243,18 @@ private HoodieLogBlock readBlock() throws IOException {
     }
   }
 
+  private Option<Schema> getTargetReaderSchemaForBlock() {
+    // we should use write schema to read log file,
+    // since when we have done some DDL operation, the readerSchema maybe different from writeSchema, avro reader will throw exception.
+    // eg: origin writeSchema is: "a String, b double" then we add a new column now the readerSchema will be: "a string, c int, b double". it's wrong to use readerSchema to read old log file.
+    // after we read those record by writeSchema,  we rewrite those record with readerSchema in AbstractHoodieLogRecordReader
+    if (internalSchema.isEmptySchema()) {
+      return Option.ofNullable(this.readerSchema);
+    } else {
+      return Option.empty();
```

Comment:
> even w/o enabling comprehensive schema evolution, one can evolve schema w/ hudi right.
So, in such a case, we will still be reading a record written in older schema w/ a newer schema. is that ok?

That's not just ok, that might be just the only way for us to read it. Idea is we only try to read in different schema if it's explicitly requested (like for ex, by schema evolution), otherwise we always read by the same schema file was written in.

> bcoz, in AbstractHoodieLogRecordReader, within composeEvolvedSchemaTransformer, we add a transformer to rewrite the record only if comprehensive schema evolution is enabled (i.e. internalSchema is non empty)

Good point. Minor evolutions (w/o comprehensive support) will be evened out when we will be writing the records back, that's the reason it's currently working as expected. We can clean this up in a follow-up

Reply:
> Minor evolution

What is the minor evolution here ?

Reply:
By minor evolutions i referred to schema evolutions that don't require enabling support for Comprehensive Schema Evolution (aka "schema.on.read")